### PR TITLE
[JENKINS-42971] CpsScmFlowDefinition sends build to SCMFileSystem in …

### PR DESCRIPTION
**Pull request by MartinKosicky**
I added in CpsScmFlowDefinition (create) to send to lightweight checkout the build value into  SCMFileSystem.of which is changed in https://github.com/jenkinsci/scm-api-plugin/pull/160. The goal is to be able to expand branch value by build params if branch name contains for example ${BRANCH}

### CODE REVIEW

**Comment by MartinKosicky** [comment]
> Ought to include an integration test picking up [jenkinsci/git-plugin#1305](https://github.com/jenkinsci/git-plugin/pull/1305) and demonstrating effectiveness in solving the stated bug.

Im planning to add integration test and unit test in Git plugin

**Comment by MartinKosicky** [comment]
> Ought to include an integration test picking up [jenkinsci/git-plugin#1305](https://github.com/jenkinsci/git-plugin/pull/1305) and demonstrating effectiveness in solving the stated bug.

Done

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:204]:
```suggestion
    @Issue("JENKINS-42971")
    @Test public void lightweight_branch_parametrised() throws Exception {
```

(also, https://en.wiktionary.org/wiki/parametrise#Alternative_forms left me rather baffled)

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:218]: I think clearer to construct a `GitSCM` directly?

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:221]:
```suggestion
```

Irrelevant here?

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:225]:
```suggestion
```

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:227]: Probably a more straightforward way to assert this. Or just delete entirely—the fact that the build ran at all (when the repo only contains the `master2` branch) is proof enough that the fix worked, right?

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (MartinKosicky) [reply src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:223]: I was wondering if the test shouldnt run faster, but ur right

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (MartinKosicky) [reply src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:220]: actually I couldnt get my head around that API how to easily create it. But i can change it somehow

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (MartinKosicky) [reply src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:229]: I wasn't sure here, because I saw other tests parsing the log file, so I thought that this would be a proof. Also I am not sure if the sample repo doesn't create master automatically now or in the future

Comment (MartinKosicky) [reply src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java]: Oki I simplified the test, and somehow created the scm directly, removed the useless log recording check, and the performence hits.

**Comment by MartinKosicky** [comment]
hmm, are these check fails related to this PR ?

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (MartinKosicky) [reply src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java]: sorry for that, its fixed

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (MartinKosicky) [reply src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java]: done

**Comment by jglick** [comment]
> are these check fails related to this PR ?

From a quick glance, looks like infrastructure problems with Windows agents; retriggered build.

**Comment by MartinKosicky** [comment]
> > are these check fails related to this PR ?
> 
> From a quick glance, looks like infrastructure problems with Windows agents; retriggered build.

Now there is a bunch of
`No test report files were found. Configuration error?`
and
`Caused: java.io.IOException: Remote call on JNLP4-connect connection from 20.102.0.182/20.102.0.182:11009 failed`

:(

**Comment by jglick** [comment]
https://ci.jenkins.io/job/Plugins/job/workflow-cps-plugin/job/PR-577/10/execution/node/53/ is https://issues.jenkins.io/browse/JENKINS-69229

**Comment by jglick** [comment]
OK, that much seems to be fixed, sorry for the interruption. @dduportal FYI

**Comment by MartinKosicky** [comment]
Ok this is testCase I was working on but not that test, hmm.  Locally they were all sucesful.  I will assume that I broke something?

File: pom.xml
Comment (jglick) [inline pom.xml:224]: also specify below for `tests` classifier

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition.java
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition.java:118]:
```suggestion
            try (SCMFileSystem fs = SCMFileSystem.of(build.getParent(), scm, null, build)) {
```

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:209]:
```suggestion
        sampleRepo.git("checkout", "-b", "master2");
```

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:214]:
```suggestion
        p.addProperty(new ParametersDefinitionProperty(new StringParameterDefinition("BRANCH", "", "")));
```
or just
```suggestion
        p.addProperty(new ParametersDefinitionProperty(new StringParameterDefinition("BRANCH", "")));
```

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:218]:
```suggestion
                Collections.singletonList(new BranchSpec("${BRANCH}")), null, null, Collections.emptyList());
```
?

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:224]:
```suggestion
        r.assertBuildStatusSuccess(p.scheduleBuild2(0, new ParametersAction(new StringParameterValue("BRANCH", "master2"))));
```

Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:224]: For reference, what is the behavior with the `git` bump locally reverted?

**Review by jglick — APPROVED** [review]

**Comment by jglick** [comment]
Edit docs: https://github.com/jenkinsci/workflow-cps-plugin/pull/329#discussion_r449219600

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (MartinKosicky) [reply src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:224]: > For reference, what is the behavior with the `git` bump locally reverted?

It fails as expected (as before)
`
22.135 [p #1] hudson.plugins.git.GitException: Command "git.exe fetch --tags --force --progress --prune -- origin +refs/heads/${BRANCH}:refs/remotes/origin/${BRANCH}" returned status code 128:
`
because the BRANCH isnt expanded. But it is backward compatible if I specify master2 directly it passes.

**Comment by MartinKosicky** [comment]
hmm

**Comment by MartinKosicky** [comment]
@jglick do you have any suggestion if I should bump the versions in git plugin? or what should we do here ?

**Comment by MartinKosicky** [comment]
When I import the incremental build of git plugin to my Idea, I see different versions than those that were specified in the incremental .pom file. hmmm

**Comment by jglick** [comment]
> When I import the incremental build of git plugin to my Idea, I see different versions

You may need to configure your IDE to honor https://github.com/jenkinsci/workflow-cps-plugin/blob/9101f1bd0b7e6f8873d96bed8e16a7505e2186b3/.mvn/maven.config#L1 NetBeans does this automatically; I have no idea about IDEA (sorry for the pun). If in doubt, command-line `mvn` is authoritative.

```
Failed to load: Git plugin (git 4.12.0-rc4821.0fa_d6b_a_a_0928)
 - Update required: Jenkins Mailer Plugin (mailer 414.vcc4c33714601) to be updated to 435.v79ef3972b_5c7 or higher
 - Update required: Token Macro Plugin (token-macro 293.v283932a_0a_b_49) to be updated to 308.v4f2b_ed62b_b_16 or higher
 - Update required: Structs Plugin (structs 318.va_f3ccb_729b_71) to be updated to 324.va_f5d6774f3a_d or higher
```

possibly just means that this should subsume #574.

**Comment by MartinKosicky** [comment]
Yes I am taking the incrementals, i saw this behaviour also in mvn dependency tree to confirm.

**Comment by jglick** [comment]
So what exactly are you seeing then?

**Comment by MartinKosicky** [comment]
> So what exactly are you seeing then?

I tried to bump the BOM that you mentioned in that PR, and it helped, I guess that was the problem

**Comment by jglick** [comment]
(draft until all deps are released)

File: pom.xml
Comment (jglick) [inline pom.xml:78]: I guess it should now suffice to pick up https://github.com/jenkinsci/bom/pull/1607 as in #634?

Comment (MarkEWaite) [reply pom.xml:78]: > I guess it should now suffice to pick up [jenkinsci/bom#1607](https://github.com/jenkinsci/bom/pull/1607) as in #634?

Yes.  For my own testing use, I created #633 as a draft pull request.  I confirmed that the originally reported issue is fixed and experimented with one branch strategy while investigating a null pointer exception in git plugin 4.14.0.  The null pointer exception is fixed in [git plugin 4.14.1](https://github.com/jenkinsci/git-plugin/releases/tag/git-4.14.1).

**Comment by MartinKosicky** [comment]
hmm the build seemed to completely change ,  now i have to figure out how to build this thing in idea :(

File: pom.xml
Comment (dwnusbaum) [inline pom.xml:47]: @MartinKosicky Sorry, but #612 turned this into a multi-module project, so you've got an unresolved merge conflict here now. I think you want to delete all of the changes in this file and apply any dependency updates to `plugin/pom.xml` instead.

Comment (MartinKosicky) [reply pom.xml]: hmm im getting "child.scm.connection.inherit.append.path is not allowed here in pom.xml"

File: pom.xml
Comment (dwnusbaum) [inline pom.xml:52]:
```suggestion
</project>

```

File: pom.xml
Comment (dwnusbaum) [reply pom.xml]: My guess is either that you need to update Maven or there is some bug with the way the IDEA tries to interpret `pom.xml` on its own now that this is a more complex project. I think there is a setting somewhere where you can tell IDEA to use Maven for the build and bypass the default behavior, but I am not 100% sure.

**Comment by MartinKosicky** [comment]
I think it shouls be ok now

**Comment by MartinKosicky** [comment]
@MarkEWaite hello, will you have time so we can finalize these 3 PRs. This one is the last and I think you also tested this already when we were working with the git-scm

**Comment by MarkEWaite** [comment]
I've added it to the [plugins.txt file](https://github.com/MarkEWaite/docker-lfs/blob/e0adf13aeca33f2d0809ad0f22e32eba357c4e0a/plugins.txt#L153) that I'm using for my [testing of git plugin 5.0.0 and git client plugin 4.0.0](https://community.jenkins.io/t/testing-git-plugin-5-0-0-and-git-client-plugin-4-0-0/5014/5).  I'll include the details of [JENKINS-42971](https://issues.jenkins.io/browse/JENKINS-42971) in my testing.

**Comment by MarkEWaite** [comment]
I've confirmed that the incremental release of this pull request resolves [JENKINS-42971](https://issues.jenkins.io/browse/JENKINS-42971).  I tested with the git plugin 5.0.0 pre-release and the git client plugin 4.0.0 pre-release.

**Review by MarkEWaite — APPROVED** [review]
Thanks!  Approved

**Comment by MartinKosicky** [comment]
> Thanks! Approved

Thank you :). Hmm just a quick question, how does the merging process work here? Who will merge this or we need more approvals?

**Review by olamy — APPROVED** [review]

**Comment by MarkEWaite** [comment]
> > Thanks! Approved
> 
> Thank you :). Hmm just a quick question, how does the merging process work here? Who will merge this or we need more approvals?

One of the [maintainers](https://github.com/orgs/jenkinsci/teams/workflow-cps-plugin-developers/members) needs to merge it.  Since two of them have approved the pull request, I believe it is ready for merge.  Once it is merged, the continuous delivery process will release a new version of the plugin.

**Comment by basil** [comment]
Note that in this project, being ready for merge is no guarantee that a pull request will be merged in a timely fashion. I routinely file PRs that are approved by maintainers with the green checkmark yet remain unmerged (and therefore unreleased) for an extended period of time.

File: plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (jglick) [inline plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:207]:
```suggestion
    public void lightweight_branch_parametrised() throws Exception {
```

File: plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (jglick) [inline plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:222]: Do you mean to also assert that it actually loaded the specified revision of `flow.groovy`, for example by checking the altered message?

**Review by jglick — APPROVED** [review]
Seems OK though I have not followed details recently. @dwnusbaum was most recently reviewing seriously; not sure if he is around to check latest changes.

**Comment by dwnusbaum** [comment]
> @dwnusbaum was most recently reviewing seriously

No, my recent review was just to try to help resolve https://github.com/jenkinsci/workflow-cps-plugin/pull/577#issuecomment-1355502611 and a bad merge related to https://github.com/jenkinsci/workflow-cps-plugin/pull/612, which is something I worked on. I have not looked at this PR beyond that.

File: plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (jglick) [reply plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:222]: FTR with `main` reverted, the failure is as expected

```
java.lang.AssertionError: 
unexpected build status; build log was:
------
Started
hudson.plugins.git.GitException: Command "git fetch --tags --force --progress --prune -- origin +refs/heads/${BRANCH}:refs/remotes/origin/${BRANCH}" returned status code 128:
stdout: 
stderr: fatal: couldn't find remote ref refs/heads/${BRANCH}

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2734)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:2111)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$500(CliGitAPIImpl.java:87)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:623)
	at jenkins.plugins.git.GitSCMFileSystem$BuilderImpl.build(GitSCMFileSystem.java:406)
	at jenkins.scm.api.SCMFileSystem.of(SCMFileSystem.java:219)
	at jenkins.scm.api.SCMFileSystem.of(SCMFileSystem.java:191)
	at jenkins.scm.api.SCMFileSystem.of(SCMFileSystem.java:174)
	at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition.create(CpsScmFlowDefinition.java:118)
	at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition.create(CpsScmFlowDefinition.java:70)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:312)
	at hudson.model.ResourceController.execute(ResourceController.java:107)
	at hudson.model.Executor.run(Executor.java:449)
Finished: FAILURE

------

Expected: is <SUCCESS>
     but: was <FAILURE>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.jvnet.hudson.test.JenkinsRule.assertBuildStatus(JenkinsRule.java:1438)
	at org.jvnet.hudson.test.JenkinsRule.assertBuildStatus(JenkinsRule.java:1444)
	at org.jvnet.hudson.test.JenkinsRule.assertBuildStatusSuccess(JenkinsRule.java:1472)
	at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinitionTest.lightweight_brach_parametrised(CpsScmFlowDefinitionTest.java:222)
```

Would just be nice to strengthen the test to prove that it was really running from the correct branch, and not merely that the build passed.

Comment (MartinKosicky) [reply plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:222]: Hmm but on 4th august you wrote that :

>  [jglick](https://github.com/jglick) [on Aug 4](https://github.com/jenkinsci/workflow-cps-plugin/pull/577#discussion_r937732930)
Probably a more straightforward way to assert this. Or just delete entirely—the fact that the build ran at all (when the repo only contains the master2 branch) is proof enough that the fix worked, right?

I was parsing the log output before

**Comment by MarkEWaite** [comment]
I found no issues related to this change in my [testing](https://community.jenkins.io/t/testing-git-plugin-5-0-0-and-git-client-plugin-4-0-0/5014/6) of git plugin 5.0.0 and git client plugin 4.0.0.  I would like to see it merged and released.  Thanks for your patience @MartinKosicky !

File: plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
Comment (jglick) [reply plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java:222]: Sigh, this is what happens when months go by between review comments. Fine enough as it is.

**Review by jglick — APPROVED** [review]

**Comment by MartinKosicky** [comment]
Thank you :)

**Comment by jglick** [comment]
Thank _you_ for driving this forward and for your patience.

**Pull request merged**
